scripts/s2n_safety_macros.py

```diff
@@ -0,0 +1,743 @@
+import textwrap
+
+copyright = """
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+"""
+
+header = copyright + """
+#pragma once
+
+/**
+ * DO NOT DIRECTLY MODIFY THIS FILE:
+ *
+ * The code in this file is generated from scripts/s2n_safety_macros.py and any modifications
+ * should be in there.
+ */
+
+#include "error/s2n_errno.h"
+#include "utils/s2n_ensure.h"
+#include "utils/s2n_result.h"
+
+/**
+ * The goal of s2n_safety is to provide helpers to perform common
+ * checks, which help with code readability.
+ */
+
+/* Success signal value for OpenSSL functions */
+#define _OSSL_SUCCESS 1
+
+"""
+
+POSIX = dict(
+    name = "POSIX",
+    is_ok = "(result) >= S2N_SUCCESS",
+    ok = "S2N_SUCCESS",
+    error = "S2N_FAILURE",
+    ret = "int",
+    ret_doc = "`int` (POSIX error signal)",
+    expect_ok = "EXPECT_SUCCESS",
+    expect_err = "EXPECT_FAILURE_WITH_ERRNO",
+)
+
+PTR = dict(
+    name = "PTR",
+    is_ok = "(result) != NULL",
+    ok = '"ok"',
+    error = "NULL",
+    ret = "const char*",
+    ret_doc = "a pointer",
+    expect_ok = "EXPECT_NOT_NULL",
+    expect_err = "EXPECT_NULL_WITH_ERRNO",
+)
+
+RESULT = dict(
+    name = "RESULT",
+    is_ok = "s2n_result_is_ok(result)",
+    ok = "S2N_RESULT_OK",
+    error = "S2N_RESULT_ERROR",
+    ret = "s2n_result",
+    ret_doc = "`S2N_RESULT`",
+    expect_ok = "EXPECT_OK",
+    expect_err = "EXPECT_ERROR_WITH_ERRNO",
+)
+
+DEFAULT = dict(
+    name = "",
+    is_ok = RESULT['is_ok'],
+    ok = RESULT['ok'],
+    error = RESULT['error'],
+    ret = RESULT['ret'],
+    expect_ok = RESULT['expect_ok'],
+    expect_err = RESULT['expect_err'],
+)
+
+# TODO add DEFAULT and remove RESULT once all PR branches are up-to-date
+CONTEXTS = [RESULT, POSIX, PTR]
+
+max_prefix_len = max(map(lambda c: len(c['name']), CONTEXTS))
+
+def cmp_check(op):
+    return '__S2N_ENSURE((a) ' + op + ' (b), {bail}(S2N_ERR_SAFETY))'
+
+    ## TODO ensure type compatibility
```

Comment:
Are there issues for these TODOs?

Reply:
I'm tracking all of the tasks in https://github.com/aws/s2n-tls/issues/2425